types/babel-core/index.d.ts

```diff
@@ -6,13 +6,26 @@
 // TypeScript Version: 2.3
 
 import * as t from 'babel-types';
-export {t as types};
+export { t as types };
 export type Node = t.Node;
 export import template = require('babel-template');
 export const version: string;
 import traverse, { Visitor } from "babel-traverse";
 export { traverse, Visitor };
 
+// A babel plugin is a simple function which must return an object matching
+// the following interface. Babel will throw if it finds unknown properties.
+// The list of allowed plugin keys is here:
+// https://github.com/babel/babel/blob/4e50b2d9d9c376cee7a2cbf56553fe5b982ea53c/packages/babel-core/src/config/option-manager.js#L71
+export interface PluginObj<S = {}> {
+    name?: string;
+    manipulateOptions?(opts: any, parserOpts: any): void;
+    pre?(this: S, state: any): void;
```

Comment:
@Kovensky Are you sure about that? For me that isn't the case with latest babel. Only inside the visitor functions `state` is equal to `this`.

```js
const babel = require("babel-core");

function plugin() {
  return {
    pre(state) {
      state.foo = "foo";
      console.log("pre:", state.foo, ", equal:", this === state);
    },
    visitor: {
      VariableDeclaration(path, state) {
        console.log("visitor:", state.foo, ", equal:", this === state);
        state.foo = "bar";
      },
    },
    post(state) {
      console.log("post:", state.foo, ", equal:", this === state);
    },
  };
}

babel.transform(`const foo = "Hello World";`, { plugins: [plugin] });

// logs:
pre: foo , equal: false
visitor: undefined , equal: true
post: foo , equal: false
```

Notice how  `state` is never set to `bar`. Now compare this to setting state on `this`:

```js
// const babel = require("babel-core");

// function plugin() {
//   return {
//     pre(state) {
//       state.foo = "foo";
//       console.log("pre:", state.foo, ", equal:", this === state);
//     },
//     visitor: {
//       VariableDeclaration(path, state) {
//         console.log("visitor:", state.foo, ", equal:", this === state);
//         state.foo = "bar";
//       },
//     },
//     post(state) {
//       console.log("post:", state.foo, ", equal:", this === state);
//     },
//   };
// }

// babel.transform(`const foo = "Hello World";`, { plugins: [plugin] });

const babel = require("babel-core");

function plugin() {
  return {
    pre(state) {
      this.foo = "foo";
      console.log("pre:", this.foo, ", equal:", this === state);
    },
    visitor: {
      VariableDeclaration(path, state) {
        console.log("visitor:", this.foo, ", equal:", this === state);
        this.foo = "bar";
      },
    },
    post(state) {
      console.log("post:", this.foo, ", equal:", this === state);
    },
  };
}

babel.transform(`const foo = "Hello World";`, { plugins: [plugin] });

// Logs:
pre: foo , equal: false
visitor: foo , equal: true
post: bar , equal: false
```

My PR fixes exactly that and makes `this` type safe by using the same interface in `pre+post` and `visitor` functions.